src/com/facebook/buck/rules/macros/ClasspathAbiMacroExpander.java

```diff
@@ -0,0 +1,160 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.rules.macros;
+
+import com.facebook.buck.jvm.core.HasClasspathEntries;
+import com.facebook.buck.jvm.core.HasJavaAbi;
+import com.facebook.buck.model.BuildTarget;
+import com.facebook.buck.model.macros.MacroException;
+import com.facebook.buck.rules.AddToRuleKey;
+import com.facebook.buck.rules.BuildRule;
+import com.facebook.buck.rules.BuildRuleResolver;
+import com.facebook.buck.rules.CellPathResolver;
+import com.facebook.buck.rules.SourcePath;
+import com.facebook.buck.rules.SourcePathResolver;
+import com.facebook.buck.rules.args.Arg;
+import com.facebook.buck.rules.args.WriteToFileArg;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Ordering;
+import java.io.File;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/**
+ * Used to expand the macro {@literal $(classpath_abi //some:target)} to the transitive abi's jars
+ * path of that target, expanding all paths to be absolute.
+ */
+public class ClasspathAbiMacroExpander extends BuildTargetMacroExpander<ClasspathAbiMacro> {
+
+  @Override
+  public Class<ClasspathAbiMacro> getInputClass() {
+    return ClasspathAbiMacro.class;
+  }
+
+  @Override
+  protected ClasspathAbiMacro parse(
+      BuildTarget target, CellPathResolver cellNames, ImmutableList<String> input)
+      throws MacroException {
+    return ClasspathAbiMacro.of(parseBuildTarget(target, cellNames, input));
+  }
+
+  private HasClasspathEntries getHasClasspathEntries(BuildRule rule) throws MacroException {
+    if (!(rule instanceof HasClasspathEntries)) {
+      throw new MacroException(
+          String.format(
+              "%s used in classpath_abi macro does not correspond to a rule with a java classpath",
+              rule.getBuildTarget()));
+    }
+    return (HasClasspathEntries) rule;
+  }
+
+  @Override
+  public Arg makeExpandToFileArg(BuildTarget target, String prefix, Arg delegate) {
+    return new ClassPathWriteToFileArg(target, prefix, delegate);
+  }
+
+  /**
+   * Get the class abi jar if present for the rule otherwise return the rule's output
+   *
+   * @param rule The rule whose jar path needs to be returned
+   * @param ruleResolver BuildRuleResolver
+   * @return class abi jar or output jar if not found
+   */
+  private SourcePath getJarPath(BuildRule rule, BuildRuleResolver ruleResolver) {
```

Comment:
This needs to be annotated with `@Nullable`.

Reply:
done